sarracen/interpolate.py

```diff
@@ -102,3 +105,108 @@ def interpolate2D(data: SarracenDataFrame,
+    :param pixcount: The number of pixels in the output over the entire cross-sectional line.
+    :return: The interpolated output, in a 1-dimensional numpy array.
+    """
+    if np.isclose(y2, y1) and np.isclose(x2, x1):
+        raise ValueError('Zero length cross section!')
+
+    if pixcount <= 0:
+        raise ValueError('pixcount must be greater than zero!')
+
+    if kernel.ndims != 2:
+        raise ValueError("Kernel must be two-dimensional!")
+
+    output = np.zeros(pixcount)
+
+    # determine the slope of the cross-section line
+    gradient = 0
+    if not np.isclose(x2, x1):
+        gradient = (y2 - y1) / (x2 - x1)
+    yint = y2 - gradient * x2
+
+    # determine the fraction of the line that one pixel represents
+    xlength = np.sqrt((x2 - x1) ** 2 + (y2 - y1) ** 2)
+    pixwidth = xlength / pixcount
+    xpixwidth = (x2 - x1) / pixcount
+
+    # copy necessary data columns into a new dataframe for vectorized operations
+    parts = pd.DataFrame()
+    parts['x'] = data[x]
+    parts['y'] = data[y]
+    parts['weight'] = data['m'] / (data['rho'] * data['h'] ** 2)
+    parts['h'] = data['h']
+    parts['term'] = parts['weight'] * data[target]
+
+    # filter out particles with 0 weight
+    parts = parts[parts['weight'] > 0]
+    parts = parts.drop('weight', axis=1)
+
+    # the intersections between the line and a particle's 'smoothing circle' are
+    # found by solving a quadratic equation with the below values of a, b, and c.
+    # if the determinant is negative, the particle does not contribute to the
+    # cross-section, and can be removed.
+    aa = 1 + gradient**2
+    parts['bb'] = 2 * gradient * (yint - parts[y]) - 2 * parts[x]
+    parts['cc'] = parts[x] ** 2 + parts[y] ** 2 - 2 * yint * parts[y] + yint ** 2 - (kernel.radkernel * parts['h']) ** 2
+    parts['det'] = parts['bb'] ** 2 - 4 * aa * parts['cc']
+    parts = parts[parts['det'] >= 0]
+    parts['det'] = np.sqrt(parts['det'])
+    parts = parts.drop(['cc'], axis=1)
+
+    # the starting and ending x coordinates of the lines intersections with a particle's smoothing circle
+    parts['xstart'] = ((-parts['bb'] - parts['det']) / (2 * aa)).clip(lower=x1, upper=x2)
+    parts['xend'] = ((-parts['bb'] + parts['det']) / (2 * aa)).clip(lower=x1, upper=x2)
+    parts = parts.drop(['det', 'bb'], axis=1)
+
+    # the start and end distances which lie within a particle's smoothing circle.
+    parts['rstart'] = np.sqrt((parts['xstart']-x1)**2 + ((gradient * parts['xstart'] + yint)-y1)**2)
+    parts['rend'] = np.sqrt((parts['xend']-x1)**2 + (((gradient * parts['xend'] + yint)-y1)**2))
+    parts = parts.drop(['xstart', 'xend'], axis=1)
+
+    # the maximum and minimum pixels that each particle contributes to.
+    parts['ipixmin'] = np.rint(parts['rstart']/pixwidth).clip(lower=0, upper=pixcount)
+    parts['ipixmax'] = np.rint(parts['rend']/pixwidth).clip(lower=0, upper=pixcount)
```

Comment:
Looking at this again, it would be worth seeing how using apply() to get ipixmin/ipixmax would compare. It would avoid the memory allocation/deallocation.

I'll send you a big data set to test with so that memory allocation/deallocation will be more noticeable.